[WIP] Initial httpd image

mnagy opened the pull request:
This is my first stab at the httpd image that serves static html pages. I've not yet looked into how to use this for our perl and php images, I will do that tomorrow. I've basically took conf files as they were in httpd24 SCL and copied them into a contrib directory and modified them. This allows us to run the container as apache user.

I didn't do any tests yet, not sure if we want to have this as a standalone image as well.

cc @mfojtik @bparees @jhadvig: You guys might want to see this.

mnagy committed:
Initial commit

mnagy committed:
Add contrib directory, direct copy from httpd24 SCL

mnagy committed:
Add run script

mnagy committed:
Apache configuration changes

mfojtik commented on 2.4/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+ENV IMAGE_TAGS              httpd,httpd24
+ENV IMAGE_EXPOSE_SERVICES   8080:http
+
+EXPOSE 8080
```
any reason why this is 8080? i thought the standard http port is 80

bparees replied:
would require root to run on 80.

bparees replied:
also we've standardized on 8080 for all our other images.

mfojtik commented on 2.4/contrib/httpd/conf/httpd.conf:
```diff
@@ -0,0 +1,353 @@
+# e-mailed.  This address appears on some server-generated pages, such
+# as error documents.  e.g. admin@your-domain.com
+#
+ServerAdmin root@localhost
```
apache@localhost

mfojtik replied:
actually this should be configurable

mfojtik commented:
This is STI enabled image and I miss .sti/bin/assemble and .sti/bin/run script (which will be the entrypoint)

mnagy committed:
STI-ize the image

We now base the image on openshift/sti-base:
 - change FROM for Dockerfile
 - add .sti/bin scripts
 - move SCL enablement script to contrib as .bashrc

mnagy committed:
Add support for templated configuration files

mnagy committed:
Add test script

mnagy committed:
Better SCL enablement and static config

The SCL enablement is now more or less bulletproof. It should work for
every bash invocation and also be totally transparent to any user.

We are now ready to be consumed by other images. They need to add their
scl enablement script into /opt/openshift/scl_enable.d/ from where it
will be read by all bash invocations. Additional httpd configs can be
added to /opt/openshift/httpd/conf.d/, ideally by the run script.
Additionally, the run script should source /opt/openshift/httpd_util.sh
and call configure_httpd() and exec_httpd().

We also add a run script that adds a simple config that sets
DocumentRoot to /opt/openshift/www/html. The assemble script will copy
the html files there.

mnagy commented:
Lot of changes. This is actually starting to look good. The scl enablement is pretty good. You only need to add a file to /opt/openshift/scl_enable.d/ and it'll get sourced in all cases, login shell or not, interactive or not, even in scripts, without you needing to do anything else.

This image can now be used as a base image for others. I already tried reworking the PHP image a bit to work with it: https://github.com/bparees/sti-php/issues/1

@mfojtik @bparees PTAL

mnagy commented:
Also, no rhel7 yet, will add that later if there won't be any bad reviews :)

Also, this has a disadvantage that the downstream images need to COPY contrib and do other stuff that is normally handled by ONBUILD instructions that are in sti-base. The reason is that with ONBUILD instructions, the image can't be used with sti to serve static html pages.

mnagy commented:
@mfojtik @bparees any comments?

mfojtik commented on 2.4/.sti/bin/assemble:
```diff
@@ -0,0 +1,6 @@
+
+set -e
+
+echo "---> Installing application source"
```
add " ..." suffix (we agreed to change it in all images, so I will be now enforcing it in all pull requests ;-)

mfojtik commented on 2.4/.sti/bin/run:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+source /opt/openshift/httpd_util.sh
```
this is not what RHCSL folks will want to see

mnagy replied:
I am aware, but I don't know about any location that I should prefer :(

mfojtik replied:
`/usr/local` was the last compromise. we should be switching to that.

mfojtik commented on 2.4/Dockerfile:
```diff
@@ -0,0 +1,31 @@
+    chown -R default.default /opt/openshift/www /var/log/httpd24 /opt/rh/httpd24/root/var/run/httpd
+
+# Make sure SCL is enabled for every interactive and/or login sessions.
+COPY scl_enable.sh /etc/profile.d/
```
is it worth to have 2 layers just to copy one file to multiple directories? can't you just symlink this?

mfojtik commented on 2.4/contrib/httpd_util.sh:
```diff
@@ -0,0 +1,14 @@
+# source
```
source what?

mfojtik replied:
I think this is pretty ugly. Requiring people that use this image as a base layer to 'source' some script is not what I think we should be doing.

The better approach would be to rename/replace the default ENTRYPOINT in the latter image (you can preserve it and call it from your ENTRYPOINT or you can just override it).

mfojtik commented on 2.4/contrib/httpd_util.sh:
```diff
@@ -0,0 +1,14 @@
+export HTTPD_SERVER_ADMIN=${HTTPD_SERVER_ADMIN:-default@localhost}
+
+function configure_httpd() {
+  echo "---> Configuring Apache"
```
add " ..." here as you want till this finish. Would be also useful to explain users what is actually being configured.

mfojtik commented on 2.4/contrib/httpd_util.sh:
```diff
@@ -0,0 +1,14 @@
+
+function configure_httpd() {
+  echo "---> Configuring Apache"
+  for template in $(find /opt/openshift/httpd/ -name '*.template'); do
```
the /opt/openshift dir is unfortunate and will change (RHCSL wants to occupy the /usr/local/etc)

mfojtik commented on 2.4/scl_enable.sh:
```diff
@@ -0,0 +1,5 @@
+for i in /opt/openshift/scl_enable.d/*.sh ; do
```
this black magic should be done elsewhere, speak to RHCSL folks

mfojtik commented on 2.4/test/html-test-app/index.html:
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
Can we generate this default page and make it show something useful? Like the container environment or help? For example "how I replace this page with my page?"

mfojtik commented:
You can copy the ONBUILD to this image if you want to use it as a 'builder' in future. That will force people to follow the 'standards' (iow. have the files in the 'known' locations'). I'm aware that this image can also be used as it is, but in that case the ONBUILD are ignored, right?

mnagy commented:
No, they are not. We were talking about that actually. If you use ONBUILDs, sti build will break for some reason. An ideal case would really be to have a base image with ONBUILDs and have an image that would only serve static html pages. Which in reality would be more useful as an httpd image usage example.

mfojtik commented:
@mnagy yeah that is in case that you want to 'build' something using this image. I think we should talk whether this will be a 'builder' image or just serve as 'base' for other builders, like PHP, Python OR 'static'.

mnagy commented:
cc @notroj @hanzz

mnagy commented:
@mfojtik I'd like this image to be a base for other images (ONBUILD instructions included) and have another one that serves static files. I wonder if that one could be included in this same repository? It'd be perfect mainly as an example of how to base other images on it.

mnagy commented:
This is ancient, closing.
